File: typescript/0020_valid_parentheses.js
/*
 * Given a string containing just the characters '(', ')', '{', '}', '[' and ']', determine if the input string is valid.
 * An input string is valid if:
 * Open brackets must be closed by the same type of brackets.
 * Open brackets must be closed in the correct order.
 * Note that an empty string is also considered valid.
*/

const hash = {
  "(": ")",
  "{": "}",
  "[": "]"
};

const validParentheses = string => {
  const stack = [];
  for (let i = 0; i < string.length; i++) {
    const current = string.charAt(i);
    if (hash.hasOwnProperty(current)) {
      stack.push(current);
    } else {
      const popped = stack.pop();
      if (hash[popped] !== current) {
        return false;
      }
    }
  }

  return stack.length === 0;
};

File: cpp/0020_valid_parentheses/0020_valid_parentheses.cc
#include "0020_valid_parentheses.h"
#include <iostream>
#include <stack>
#include <string>
#include <unordered_map>

using std::stack;
using std::unordered_map;

bool Solution::isValid(string s) {
  unordered_map<char, char> dict({{'(', ')'}, {'[', ']'}, {'{', '}'}});
  stack<char> stack;

  for (const char &letter : s) {
    if (dict.find(letter) != dict.end()) {
      stack.push(dict[letter]);
      continue;
    }

    if (stack.empty()) {
      return false;
    }

    const char &popped = stack.top();

    if (letter != popped) {
      return false;
    }

    stack.pop();
  }

  return stack.empty();
}
